src/FileStorageBase/Network/Address.cs

```diff
@@ -0,0 +1,246 @@
+        }
+
+        public static Address FromString(string s)
+        {
+            Multiaddress m;
+            try
+            {
+                m = Multiaddress.Decode(s);
+            }
+            catch (Exception)
+            {
+                m = Multiaddress.Decode(MultiAddrStringFromHostAddr(s));
+            }
+            return new Address(m);
+        }
+
+        public bool Equals(Address other)
+        {
+            if (other is null) return false;
+            return other.ma.Equals(ma);
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj is null) return false;
+            if (ReferenceEquals(this, obj)) return true;
+            if (obj is Address addr)
+                return Equals(addr);
+            return false;
+        }
+
+        public override int GetHashCode() => ma.ToString().GetHashCode();
+
+        /// <summary>
+        /// multiaddrStringFromHostAddr converts "localhost:8080" to "/dns4/localhost/tcp/8080"
+        /// </summary>
+        /// <param name="host"></param>
+        /// <returns></returns>
+        private static string MultiAddrStringFromHostAddr(string host)
+        {
+            if (0 < host.Length && host[0] == ':') host = "0.0.0.0" + host;
+            if (host.Last() == ':') host += "0";
+            var endPoint = IPEndPoint.Parse(host);
+            var addr = endPoint.Address;
+            var port = endPoint.Port;
+            var prefix = "/dns4";
+            var s = addr.ToString();
+            if (addr.AddressFamily == AddressFamily.InterNetwork)
+                prefix = "/ip4";
+            else if (addr.AddressFamily == AddressFamily.InterNetworkV6)
+                prefix = "/ip6";
+            return string.Join('/', prefix, s, L4Protocol, port);
+        }
+
+        /// <summary>
+        /// IPAddrFromMultiaddr converts "/dns4/localhost/tcp/8080" to "192.168.0.1:8080".
+        /// </summary>
+        /// <param name="multiaddr"></param>
+        /// <returns></returns>
+        public static string IPAddrFromMultiaddr(string multiaddr)
+        {
+            return FromString(multiaddr).ToIPAddressString();
+        }
+
+        public override string ToString()
+        {
+            return ma.ToString();
+        }
+
+        private string DialArgs(Multiaddress ma)
+        {
+            var res = DialArgsComponents(ma);
+            if (res.HostName)
+            {
+                return res.Network switch
+                {
+                    "ip" or "ip4" or "ip6" => res.IP,
+                    "tcp" or "tcp4" or "tcp6" or "udp" or "udp4" or "udp6" => res.IP + ":" + res.Port,
+                    _ => throw new InvalidOperationException("unreachable"),
+                };
+            }
+            return res.Network switch
+            {
+                "ip4" => res.IP,
+                "tcp4" or "udp4" => res.IP + ":" + res.Port,
+                "tcp6" or "udp6" => "[" + res.IP + "]" + ":" + res.Port,
```

Comment:
this doesnt work for https://github.com/neo-project/neo-modules/blob/f6b981a30e42b78df261a1c33cbcb02be5fd1a1d/src/FileStorageBase/Cache/ClientCache/MultiClient.cs#L49

Reply:
I will add more validation check.

Reply:
Done.